content/file/file.go

```diff
@@ -0,0 +1,542 @@
+		return &buffer
+	},
+}
+
+const (
+	// AnnotationDigest is the annotation key for the digest of the uncompressed content.
+	AnnotationDigest = "io.deis.oras.content.digest"
+	// AnnotationUnpack is the annotation key for indication of unpacking.
+	AnnotationUnpack = "io.deis.oras.content.unpack"
+	// defaultBlobMediaType specifies the default blob media type.
+	defaultBlobMediaType = ocispec.MediaTypeImageLayer
+	// defaultBlobDirMediaType specifies the default blob directory media type.
+	defaultBlobDirMediaType = ocispec.MediaTypeImageLayerGzip
+	// defaultSizeLimit specifies the default size limit for pushing no-name contents.
+	defaultSizeLimit = 1 << 22 // 4 MiB
+)
+
+// Store represents a file system based store, which implements `oras.Target`.
+type Store struct {
+	TarReproducible           bool
+	AllowPathTraversalOnWrite bool
+	DisableOverwrite          bool
+
+	workingDir   string
+	digestToPath sync.Map // map[digest.Digest]string
+	nameToStatus sync.Map // map[string]chan struct{}
+	tmpFiles     sync.Map // map[string]bool
+
+	fallbackStorage content.Storage
+	resolver        content.TagResolver
+	graph           *graph.Memory
+}
+
+// New creates a file store, using a default limited memory storage
+// as the fallback storage for contents without names.
+// When pushing content without names, the size of content being pushed
+// cannot exceed the default size limit: 4 MiB.
+func New(workingDir string) *Store {
+	return NewWithFallbackLimit(workingDir, defaultSizeLimit)
+}
+
+// NewWithFallbackLimit creates a file store, using a default
+// limited memory storage as the fallback storage for contents without names.
+// When pushing content without names, the size of content being pushed
+// cannot exceed the size limit specified by the `limit` parameter.
+func NewWithFallbackLimit(workingDir string, limit int64) *Store {
+	m := cas.NewMemory()
+	ls := content.LimitStorage(m, limit)
+	return NewWithFallbackStorage(workingDir, ls)
+}
+
+// NewWithFallbackStorage creates a file store,
+// using the provided fallback storage for contents without names.
+func NewWithFallbackStorage(workingDir string, fallbackStorage content.Storage) *Store {
+	return &Store{
+		workingDir:      workingDir,
+		fallbackStorage: fallbackStorage,
+		resolver:        resolver.NewMemory(),
+		graph:           graph.NewMemory(),
+	}
+}
+
+// Close cleans up all the temp files used by the file store.
+func (s *Store) Close() error {
+	var errs []string
+	s.tmpFiles.Range(func(name, _ interface{}) bool {
+		if err := os.Remove(name.(string)); err != nil {
+			errs = append(errs, err.Error())
+		}
+		return true
+	})
+	return errors.New(strings.Join(errs, "; "))
+}
+
+// Fetch fetches the content identified by the descriptor.
+// If name is not specified in the descriptor,
+// the content will be fetched from the fallback storage.
+func (s *Store) Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error) {
```

Comment:
What if there is a content described by two different names? Suppose one is called `foo.txt`, and another is called `bar.txt`, their digest are the same.
```go
desc_foo := ocispec.Descriptor{
                MediaType: "test",
		Digest:    digest.FromBytes(content),
		Size:      int64(len(content)),
		Annotations: map[string]string{
			ocispec.AnnotationTitle: "foo.txt",
		},
}
desc_bar := ocispec.Descriptor{
                MediaType: "test",
		Digest:    digest.FromBytes(content),
		Size:      int64(len(content)),
		Annotations: map[string]string{
			ocispec.AnnotationTitle: "bar.txt",
		},
}
```
The `foo.txt` has been pushed to the file store, while `bar.txt` is not.
In this case, `Exists(ctx, desc_bar)` would return false since `bar.txt` is not there.
For `bar.txt`, if we try to fetch the content from the current storage first, we would get the content successfully, as the content is the same as `foo.txt`.  Would this behavior be confusing?

Reply:
Consider `foo` exists in the store but `bar` does not.

Assume

```go
desc := desc_foo
desc.Annotations = nil
```

- `Exists(ctx, desc_foo)` returns true.
- `Exists(ctx, desc_bar)` returns false.
- `Exists(ctx, desc)` returns true.
- `Fetch(ctx, desc_foo)` returns true.
- `Fetch(ctx, desc_bar)` returns false.
- `Fetch(ctx, desc)` returns true.

When the annotation presents, we validate the digest and the annotation. Otherwise, we just validate the digest.

Reply:
This makes sense.